File: Data Structures & Algorithms/islands-and-treasure/submission-6.java
class Solution {

    int[][] dirs = {{-1,0}, {1,0}, {0,-1}, {0,1}};
    public void islandsAndTreasure(int[][] grid) {
        int rows = grid.length, cols = grid[0].length;
        Queue<int[]> q = new LinkedList<>();

        for(int r=0; r<rows; r++){
            for(int c=0; c<cols; c++){
                if(grid[r][c] == 0){
                    q.add(new int[]{r,c});
                }
            }
        }

        int res = 1;
        while(!q.isEmpty()){
            int size = q.size();
            for(int i=0; i<size; i++){
                int[] crr = q.poll();

                int r = crr[0];
                int c = crr[1];

                for(int[] dir:dirs){
                    int nr = dir[0] + r;
                    int nc = dir[1] + c;

                    if(nr >= 0 && nc >= 0 && nr < grid.length && nc < grid[0].length && grid[nr][nc] == Integer.MAX_VALUE){
                        grid[nr][nc] = res;
                        q.add(new int[]{nr, nc});
                    }
                }
            }

            res++;
        }
    }
}

File: Data Structures & Algorithms/rotting-fruit/submission-6.py
class Solution:
    def isValid(self, r, c, rows, cols):
        if r >= 0 and  c>=0 and r < rows and c < cols:
            return True
        return False

    def orangesRotting(self, grid: List[List[int]]) -> int:
        rows,cols = len(grid), len(grid[0])
        q = deque()
        dirs = [[-1,0], [1,0], [0,1], [0,-1]]
        
        fresh = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    q.append((r,c))
                if grid[r][c] == 1:
                    fresh +=1
        
        res = 0
        while fresh > 0 and q:
            size = len(q)

            for i in range(size):
                r,c = q.popleft()

                for dir in dirs:
                    nr = r + dir[0]
                    nc = c + dir[1]
                    if self.isValid(nr, nc, rows, cols) and grid[nr][nc] == 1:
                        fresh -=1
                        grid[nr][nc] = 2
                        q.append((nr,nc))

            res += 1
        return res if fresh == 0 else -1

